File: app/lib/domain/Topic.ts
export interface TopicJSON {
  name: string;
  type: TopicTypes;
}

export enum TopicTypes {
  DataStructures = "DataStructures",
  Algorithms = "Algorithms",
  Concepts = "Concepts",
  Miscellaneous = "Miscellaneous",
}

export class Topic {
  public constructor(
    private readonly _name: string,
    private readonly _type: TopicTypes
  ) {}

  // ====================
  // Public API
  // ====================

  public get name(): string {
    return this._name;
  }

  public get type(): TopicTypes {
    return this._type;
  }

  // ====================
  // Domain Methods
  // ====================

  public toJSON(): TopicJSON {
    return {
      name: this._name,
      type: this._type,
    };
  }

  static fromJSON(json: TopicJSON): Topic {
    if (!json?.name || json.type === undefined) {
      throw new Error("Invalid Topic JSON");
    }
    return new Topic(json.name, json.type);
  }

  public toString(): string {
    return `Topic(name:${this._name}, type:${this._type})`;
  }
}

File: backend/app/core/domain/Topic.py
from app.api.schemas.schemas import TopicJSON, TopicTypes

class Topic:
    def __init__(self, name: str, topic_type: TopicTypes):
        self._name = name
        self._type = topic_type

    @property
    def name(self) -> str:
        return self._name

    @property
    def type(self) -> TopicTypes:
        return self._type

    def to_json(self) -> TopicJSON:
        return TopicJSON(
            name=self._name,
            type=self._type.value,
        )

    @classmethod
    def from_json(self, json: TopicJSON) -> 'Topic':
        if not json.name or not json.type:
            raise ValueError("Invalid Topic JSON")
        return self(json.name, TopicTypes(json.type))